### PR TITLE
Add ability to send many location checks in one go

### PR DESCRIPTION
**Pull request by Jarno458**

### CODE REVIEW

File: Archipelago.cpp [hunk 1]
```diff
@@ -241,34 +241,46 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
 }
 
 void AP_SendItem(int64_t idx) {
-    printf("AP: Checked '%s'. Informing Archipelago...\n", getLocationName(ap_game,idx).c_str());
+    AP_SendItem(std::vector<int64_t>{ idx });
+}
+void AP_SendItem(std::vector<int64_t> const& locations) {
+    printf("AP: ");
```
Comment (N00byKing) [inline Archipelago.cpp:247]: Put this in the print below

File: Archipelago.cpp [hunk 2]
```diff
@@ -241,34 +241,46 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+    for (int64_t idx : locations) {
+        printf("Checked '%s'.", getLocationName(ap_game, idx).c_str());
```
Comment (N00byKing) [inline Archipelago.cpp:249]: Add newline here

File: Archipelago.cpp [hunk 3]
```diff
@@ -241,34 +241,46 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+    }
+    printf("Informing Archipelago...\n", locations.size());
```
Comment (N00byKing) [inline Archipelago.cpp:251]: `locations.size` unused? This print line can be removed entirely though, doesnt really add anything

File: Archipelago.cpp [hunk 4]
```diff
@@ -241,34 +241,46 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+        };
         APSend(writer.write(req_t));
     } else {
-        for (auto itr : sp_save_root["checked_locations"]) {
-            if (itr.asInt64() == idx) {
-                return;
+        for (int64_t idx : locations) {
+            for (auto itr : sp_save_root["checked_locations"]) {
+                if (itr.asInt64() == idx) {
+                    return;
```
Comment (N00byKing) [inline Archipelago.cpp:264]: Logic here seems broken. It reads to me as: If any location was checked before (in `sp_save_root`), do not update any.
It should probably instead skip those that are checked, and only consider the new ones.
You'll need to remove the entries in `sp_save_root` from `locations` and use that to create the fake message

File: Archipelago.cpp [hunk 5]
```diff
@@ -241,34 +241,46 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+            int64_t recv_item_id = sp_ap_root["location_to_item"].get(std::to_string(idx), 0).asInt64();
+            if (recv_item_id == 0) return;
+            Json::Value fake_msg;
+            fake_msg[0]["cmd"] = "ReceivedItems";
+            fake_msg[0]["index"] = last_item_idx+1;
+            fake_msg[0]["items"][0]["item"] = recv_item_id;
+            fake_msg[0]["items"][0]["location"] = idx;
+            fake_msg[0]["items"][0]["player"] = ap_player_id;
+            std::string req;
+            parse_response(writer.write(fake_msg), req);
```
Comment (N00byKing) [inline Archipelago.cpp:276]: You are creating a new fake message object for each item. Instead, just like in the multiworld case, you need to create a message that includes all of the items at once.
Same for the location checks below

File: Archipelago.h [hunk 6]
```diff
@@ -71,7 +71,8 @@ void AP_SetLocationInfoCallback(void (*f_locrecv)(std::vector<AP_NetworkItem>));
 /* Game Management Functions */
 
 // Sends LocationCheck for given index
-void AP_SendItem(int64_t);
+void AP_SendItem(int64_t location);
+void AP_SendItem(std::vector<int64_t> const& locations);
```
Comment (N00byKing) [inline Archipelago.h:75]: Should be a set, not vector

**Review by N00byKing — CHANGES_REQUESTED** [review]
No issues with the idea, but needs a few changes.

File: Archipelago.cpp [hunk 7]
```diff
@@ -241,34 +241,49 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+        req_t[0]["locations"] = Json::arrayValue;
+        for (int64_t loc : locations) {
+            req_t[0]["locations"].append(loc);
+        };
         APSend(writer.write(req_t));
     } else {
-        for (auto itr : sp_save_root["checked_locations"]) {
-            if (itr.asInt64() == idx) {
-                return;
+        std::vector<int64_t> new_locations(locations.size());
```
Comment (N00byKing) [inline Archipelago.cpp:259]: This should also be a set. And why did you give a size?

File: Archipelago.cpp [hunk 8]
```diff
@@ -241,34 +241,49 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
...
+        for (int64_t idx : locations) {
+            for (auto itr : sp_save_root["checked_locations"]) {
+                if (itr.asInt64() == idx) {
```
Comment (N00byKing) [inline Archipelago.cpp:262]: Logic here is still broken. Consider one iteration: Have a location idx to check, but you compare against only one entry in checked_locations which, if it succeeds, you add to new_locations. Then you continue checking every other entry in checked_locations which does not make sense.

File: Archipelago.cpp (hunk 5 above)
Comment (N00byKing) [reply Archipelago.cpp:279]: Still an issue

File: Archipelago.cpp [hunk 9]
```diff
@@ -241,34 +241,49 @@ void AP_SetClientVersion(AP_NetworkVersion* version) {
 }
 
 void AP_SendItem(int64_t idx) {
-    printf("AP: Checked '%s'. Informing Archipelago...\n", getLocationName(ap_game,idx).c_str());
+    AP_SendItem(std::set<int64_t>{ idx });
+}
+void AP_SendItem(std::set<int64_t> const& locations) {
+    for (int64_t idx : locations) {
+        printf("AP: Checked '%s'.", getLocationName(ap_game, idx).c_str());
```
Comment (N00byKing) [inline Archipelago.cpp:248]: Missing newline

**Review by N00byKing — CHANGES_REQUESTED** [review]

File: Archipelago.cpp (hunk 7 above)
Comment (Jarno458) [reply Archipelago.cpp:259]: For performance, i think inserts into a set will allocate memory when needed, where i can pre-claim enough space

**Pull request merged**